auth0/src/main/java/com/auth0/android/authentication/storage/CredentialsManager.java

```diff
@@ -64,27 +66,43 @@ public void saveCredentials(@NonNull Credentials credentials) {
      * @param callback the callback that will receive a valid {@link Credentials} or the {@link CredentialsManagerException}.
      */
     public void getCredentials(@NonNull final BaseCallback<Credentials, CredentialsManagerException> callback) {
+        getCredentials(null, callback);
+    }
+
+    /**
+     * Retrieves the credentials from the storage and refresh them if they have already expired.
+     * It will fail with {@link CredentialsManagerException} if the saved access_token or id_token is null,
+     * or if the tokens have already expired and the refresh_token is null.
+     *
+     * @param callback the callback that will receive a valid {@link Credentials} or the {@link CredentialsManagerException}.
+     */
+    public void getCredentials(@Nullable final String scope,
+                               @NonNull final BaseCallback<Credentials, CredentialsManagerException> callback) {
         String accessToken = storage.retrieveString(KEY_ACCESS_TOKEN);
         final String refreshToken = storage.retrieveString(KEY_REFRESH_TOKEN);
         String idToken = storage.retrieveString(KEY_ID_TOKEN);
         String tokenType = storage.retrieveString(KEY_TOKEN_TYPE);
         Long expiresAt = storage.retrieveLong(KEY_EXPIRES_AT);
-        String scope = storage.retrieveString(KEY_SCOPE);
+        String storedScope = storage.retrieveString(KEY_SCOPE);
 
         if (isEmpty(accessToken) && isEmpty(idToken) || expiresAt == null) {
             callback.onFailure(new CredentialsManagerException("No Credentials were previously set."));
             return;
         }
         if (expiresAt > getCurrentTimeInMillis()) {
-            callback.onSuccess(recreateCredentials(idToken, accessToken, tokenType, refreshToken, new Date(expiresAt), scope));
+            callback.onSuccess(recreateCredentials(idToken, accessToken, tokenType, refreshToken, new Date(expiresAt), storedScope));
```

Comment:
if the scope asked for in this introduced method call cannot be satisfied with the existing token, this is going to deliver an invalid token back to the user in the onSuccess call. Similarly, if the refresh token originally obtained was not granted those scopes before, trying to refresh it with the new requested set of scopes will also fail.

Maybe the safest thing to do is split the scope strin (which is going to contain scopes separated by a space.) Then check if ALL of the new requested scopes are included in the original scope string, in any order. If at least one of the new requested scopes is not present on the original scope string, then throw a runtime error such as `InvalidArgumentException` as the dev made the mistake of asking a scope he knows it can't be provided.

Reply:
Interesting. But isn't there always the case in the initial token request that you are granted different scopes than you request? In this case, no error is thrown, so I don't think there should be an error thrown here.

Reply:
> But isn't there always the case in the initial token request that you are granted different scopes than you request?
You get the scopes you ask for. Can you explain your rationale here?